src/main/java/dev/morling/onebrc/CalculateAverage_royvanrijn.java

```diff
@@ -164,127 +168,161 @@ private static double round(double value) {
+                    // temperature, you're not temporary my friend
+                    final int temp = extractTemp(decimalSepPos, invNumberBytes, numberBytes);
+                    // No differences, same entry:
                     tableEntry.updateWith(temp);
                     break;
                 }
-                // Move to the next index
+                // Move to the next in the table, linear probing:
                 index = (index + 1) & TABLE_MASK;
             }
         }
         return table;
     }
 
-    private static int extractTemp(final long numberBits, final long invNumberBits, final int decimalSepPos) {
+    /*
+     * `___` ___ ___ _ ___` ` ___ ` _ ` _ ` _` ___
+     * / ` \| _ \ __| \| \ \ / /_\ | | | | | | __|
+     * | () | _ / __|| . |\ V / _ \| |_| |_| | ._|
+     * \___/|_| |___|_|\_| \_/_/ \_\___|\___/|___|
+     * ---------------- BETTER SOFTWARE, FASTER --
+     *
+     * https://www.openvalue.eu/
+     *
+     * Made you look.
+     *
+     */
+
+    private static final long DOT_BITS = 0x10101000;
+    private static final long MAGIC_MULTIPLIER = (100 * 0x1000000 + 10 * 0x10000 + 1);
+
+    private static int extractTemp(final int decimalSepPos, final long invNumberBits, final long numberBits) {
+        // Awesome idea of merykitty:
+        int min28 = (28 - decimalSepPos);
+        // Calculates the sign
         final long signed = (invNumberBits << 59) >> 63;
         final long minusFilter = ~(signed & 0xFF);
-        final long digits = ((numberBits & minusFilter) << (28 - decimalSepPos)) & 0x0F000F0F00L;
-        final long absValue = ((digits * MAGIC_MULTIPLIER) >>> 32) & 0x3FF; // filter just the result
+        // Use the pre-calculated decimal position to adjust the values
+        final long digits = ((numberBits & minusFilter) << min28) & 0x0F000F0F00L;
+        // Multiply by a magic (100 * 0x1000000 + 10 * 0x10000 + 1), to get the result
+        final long absValue = ((digits * MAGIC_MULTIPLIER) >>> 32) & 0x3FF;
+        // And perform abs()
         final int temp = (int) ((absValue + signed) ^ signed); // non-patented method of doing the same trick
         return temp;
     }
 
+    private static final long SEPARATOR_PATTERN = 0x3B3B3B3B3B3B3B3BL;
+
+    // Takes a long and finds the bytes where this exact pattern is present.
+    // Cool bit manipulation technique: SWAR (SIMD as a Register).
     private static long getDelimiterMask(final long word) {
-        long match = word ^ SEPARATOR_PATTERN;
-        return (match - 0x0101010101010101L) & ~match & 0x8080808080808080L;
+        final long match = word ^ SEPARATOR_PATTERN;
+        return (match - 0x0101010101010101L) & (~match & 0x8080808080808080L);
+        // I've put some brackets separating the first and second part, this is faster.
+        // Now they run simultaneous after 'match' is altered, instead of waiting on each other.
```

Comment:
Interesting. Out of curiosity, how do you know that?

Reply:
Running it with and without the brackets, at this point I'm just changing things and hoping for the best.

It does make sense though, first calculate (match - something) and (~match & something) without a dependency, than AND together.

Reply:
just checking if i understand this correctly... there is no parallelisation here right? the reason (1 - 2) & (~3 &4) is faster is because things end up in correct registers so the last & is faster?

Reply:
Roy can correct me, but I think he means that It's parallel within the CPU thanks to its superscalar hardware. Even a single core has multiple ALUs.

Reply:
thanks, in that case its indeed surprising that this is not done on the compiler level, pretty easy to detect...